Implement lint for obligations broken by never type fallback change

WaffleLapkin opened the pull request:
This is the second (and probably last major?) lint required for the never type fallback change.

The idea is to check if the code errors with `fallback = ()` and if it errors with `fallback = !` and if it went from "ok" to "error", lint.

I'm not happy with the diagnostic, ideally we'd highlight what bound is the problem. But I'm really unsure how to do that  (cc @jackh726, iirc you had some ideas?)

r? @compiler-errors 

Thanks @BoxyUwU with helping with trait solver stuff when I was implementing the initial version of this lint.

Tracking:
- https://github.com/rust-lang/rust/issues/123748

WaffleLapkin committed:
Implement lint for obligations broken by never type fallback change

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9149858832/job/25153958731) [(plain)](https://github.com/rust-lang/rust/commit/77da176f223b2c8032a054eaa251880f33eb47d9/checks/25153958731/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.9s done
#16 DONE 34.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

115 LL |     reuse to_reuse::opaque_arg;
116    |                     ^^^^^^^^^^
117 
+ warning: this function depends on never type fallback being `()`
+    |
+    |
+ LL |         fn opaque_ret() -> impl Trait { unimplemented!() }
+    |
+    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+    = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
+    = help: specify the types explicitly
+    = help: specify the types explicitly
+    = note: `#[warn(dependency_on_unit_never_type_fallback)]` on by default
+ 
118 error[E0391]: cycle detected when computing type of `opaque::<impl at $DIR/not-supported.rs:82:5: 82:24>::{synthetic#0}`
120    |

134    |     ^^^^^^^^^^^^^^^^^^^
134    |     ^^^^^^^^^^^^^^^^^^^
135    = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
136 
+ warning: this function depends on never type fallback being `()`
+    |
+    |
+ LL |         pub fn opaque_ret() -> impl Trait { unimplemented!() }
+    |
+    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+    = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
+    = help: specify the types explicitly
+    = help: specify the types explicitly
+ 
137 error[E0391]: cycle detected when computing type of `opaque::<impl at $DIR/not-supported.rs:85:5: 85:25>::{synthetic#0}`
139    |

162 LL |     reuse to_reuse1::foo;
163    |                      ^^^
---
To only update this specific test, also pass `--test-args delegation/not-supported.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/delegation/not-supported.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/delegation/not-supported" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/delegation/not-supported/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error: delegation with early bound generics is not supported yet
   |
   |
LL |         fn bar(&self, x: T) -> T { x }
   |         ------------------------ callee defined here
LL |         reuse GenericTrait::bar;
   |                             ^^^


error: delegation with early bound generics is not supported yet
   |
LL |         fn bar1() {}
   |         --------- callee defined here
...
...
LL |         reuse GenericTrait::bar1;
   |                             ^^^^

error: delegation with early bound generics is not supported yet
   |
   |
LL |         fn bar(&self, x: T) -> T { x }
   |         ------------------------ callee defined here
...
LL |         reuse <F as GenericTrait<T>>::bar { &self.0 }


error: delegation with early bound generics is not supported yet
   |
LL |         fn bar1() {}
   |         --------- callee defined here
...
...
LL |         reuse GenericTrait::<T>::bar1;
   |                                  ^^^^

error: delegation with early bound generics is not supported yet
   |
   |
LL |         fn bar(&self, x: T) -> T { x }
   |         ------------------------ callee defined here
...
LL |         reuse GenericTrait::bar { &F }


error: delegation with early bound generics is not supported yet
   |
LL |         fn bar1() {}
   |         --------- callee defined here
...
...
LL |         reuse GenericTrait::bar1;
   |                             ^^^^

error: delegation with early bound generics is not supported yet
   |
   |
LL |         fn foo(&self, x: i32) -> i32 { x }
   |         ---------------------------- callee defined here
LL |         reuse Trait::foo;
   |                      ^^^


error[E0049]: method `foo2` has 0 type parameters but its trait declaration has 1 type parameter
   |
   |
LL |         fn foo2<T>(&self, x: T) -> T { x }
   |                 - expected 1 type parameter
LL |         reuse Trait::foo2 { &self.0 }
   |                      ^^^^ found 0 type parameters


error: delegation with early bound generics is not supported yet
   |
   |
LL |         fn foo3<'a: 'a>(_: &'a u32) {}
   |         --------------------------- callee defined here
...
LL |         reuse <F as Trait>::foo3;


error[E0195]: lifetime parameters or bounds on method `foo3` do not match the trait declaration
   |
   |
LL |         fn foo3<'a: 'a>(_: &'a u32) {}
   |                -------- lifetimes in impl do not match this method in trait
...
LL |         reuse <F as Trait>::foo3;
   |                             ^^^^ lifetimes do not match method in trait

error: delegation with early bound generics is not supported yet
   |
   |
LL |         fn foo(&self, x: i32) -> i32 { x }
   |         ---------------------------- callee defined here
LL |         reuse Trait::foo { &self.0 }
   |                      ^^^


error: delegation with early bound generics is not supported yet
   |
   |
LL |         fn foo2<T>(&self, x: T) -> T { x }
   |         ---------------------------- callee defined here
LL |         reuse Trait::foo2 { &self.0 }
   |                      ^^^^


error: delegation with early bound generics is not supported yet
   |
   |
LL |         pub fn opaque_arg(_: impl Trait) -> i32 { 0 }
   |         --------------------------------------- callee defined here
LL |     reuse to_reuse::opaque_arg;
   |                     ^^^^^^^^^^


warning: this function depends on never type fallback being `()`
   |
   |
LL |         fn opaque_ret() -> impl Trait { unimplemented!() }
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
   = help: specify the types explicitly
   = help: specify the types explicitly
   = note: `#[warn(dependency_on_unit_never_type_fallback)]` on by default

error[E0391]: cycle detected when computing type of `opaque::<impl at /checkout/tests/ui/delegation/not-supported.rs:82:5: 82:24>::{synthetic#0}`
   |
   |
LL |         reuse to_reuse::opaque_ret; //~ ERROR cycle detected when computing type
   |
   |
note: ...which requires comparing an impl and trait method signature, inferring any hidden `impl Trait` types in the process...
   |
   |
LL |         reuse to_reuse::opaque_ret; //~ ERROR cycle detected when computing type
   |                         ^^^^^^^^^^
   = note: ...which again requires computing type of `opaque::<impl at /checkout/tests/ui/delegation/not-supported.rs:82:5: 82:24>::{synthetic#0}`, completing the cycle
note: cycle used when checking that `opaque::<impl at /checkout/tests/ui/delegation/not-supported.rs:82:5: 82:24>` is well-formed
   |
   |
LL |     impl ToReuse for u8 {
   |     ^^^^^^^^^^^^^^^^^^^
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

warning: this function depends on never type fallback being `()`
   |
   |
LL |         pub fn opaque_ret() -> impl Trait { unimplemented!() }
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
   = help: specify the types explicitly
   = help: specify the types explicitly

error[E0391]: cycle detected when computing type of `opaque::<impl at /checkout/tests/ui/delegation/not-supported.rs:85:5: 85:25>::{synthetic#0}`
   |
   |
LL |         reuse ToReuse::opaque_ret; //~ ERROR cycle detected when computing type
   |
   |
note: ...which requires comparing an impl and trait method signature, inferring any hidden `impl Trait` types in the process...
   |
   |
LL |         reuse ToReuse::opaque_ret; //~ ERROR cycle detected when computing type
   |                        ^^^^^^^^^^
   = note: ...which again requires computing type of `opaque::<impl at /checkout/tests/ui/delegation/not-supported.rs:85:5: 85:25>::{synthetic#0}`, completing the cycle
note: cycle used when checking that `opaque::<impl at /checkout/tests/ui/delegation/not-supported.rs:85:5: 85:25>` is well-formed
   |
   |
LL |     impl ToReuse for u16 {
   |     ^^^^^^^^^^^^^^^^^^^^
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
error: recursive delegation is not supported yet
##[error]  --> /checkout/tests/ui/delegation/not-supported.rs:99:22
   |
LL |         pub reuse to_reuse2::foo;
```

</details>

compiler-errors commented on compiler/rustc_hir_typeck/src/fallback.rs:
```diff
@@ -464,6 +472,46 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
+
+        // If we have no errors with `fallback = ()`, but *do* have errors with `fallback = !`,
+        // then this code will be broken by the never type fallback change.qba
+        let unit_errors = try_out(self.tcx.types.unit);
```
`try_out` is kinda a confusing name -- can you name it something like `remaining_errors_if_fallback_to` or something?

WaffleLapkin replied:
`remainign_errors_if_fallback_to` seems a bit weird, but it's better than `try_out` for sure, so I changed it.

jackh726 commented:
> I'm not happy with the diagnostic, ideally we'd highlight what bound is the problem. But I'm really unsure how to do that

So, yeah, we talked about this. IIRC, what I said was basically like "we can't find diff two error sets" but as long as we only try to report `!` fallback errors if `()` fallback errors are empty, that's probably fine...

We can chat on zulip about details here, but I would probably approach this by selecting a couple test cases where you want better diagnostics, then look at the errors that you get and what info is there, then figure out how to extract out that info. Then you can generalize across tests.

jackh726 commented on compiler/rustc_hir_typeck/src/fallback.rs:
```diff
@@ -464,6 +472,46 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
+
+        // If we have no errors with `fallback = ()`, but *do* have errors with `fallback = !`,
+        // then this code will be broken by the never type fallback change.qba
+        let unit_errors = try_out(self.tcx.types.unit);
```
So, I have to go back and look at this code path, but this is likely going to be way too slow perf.

First, we're going to end up duplicating the obligation processing, which alone is problematic.

Second, we likely only want to try to lint about fallback if there *was* any fallback (if there is no fallback, then you would expect there to be no change in errors if you change the fallback type).

compiler-errors replied:
> First, we're going to end up duplicating the obligation processing, which alone is problematic.

It really depends on how many obligations remain stalled after fallback. I would expect there are not many. I personally wouldn't particularly worry about perf here.

WaffleLapkin replied:
I could add a fast-exit for cases where `diverging_vids.is_empty()` (i.e. the only case where fallback won't happen), but as compiler-errors said, I don't think this will be perf sensitive.

compiler-errors replied:
A fast exit will actually be beneficial regardless if it's perf-sensitive. I'd rather you add that since it's not too difficult.

WaffleLapkin replied:
fast exit added 👍🏻

compiler-errors commented on compiler/rustc_hir_typeck/src/fallback.rs:
```diff
@@ -8,12 +8,17 @@ use rustc_data_structures::{
 use rustc_span::DUMMY_SP;
 use rustc_span::{def_id::LocalDefId, Span};
+use rustc_trait_selection::traits::ObligationCtxt;
+use rustc_type_ir::TyVid;
```
Don't import this. use `ty::TyVid`

compiler-errors commented on compiler/rustc_hir_typeck/src/fallback.rs:
```diff
@@ -8,12 +8,17 @@ use rustc_data_structures::{
 use rustc_hir::HirId;
-use rustc_infer::infer::{DefineOpaqueTypes, InferOk};
 use rustc_middle::bug;
+use rustc_infer::{
+    infer::{DefineOpaqueTypes, InferOk},
+    traits::ObligationCause,
+};
```
nit: Can you split this up so it's not nested? Just two `use` lines?

Also, you can probably import `ObligationCause` from the same `use` statement as `ObligationCtxt` below.

compiler-errors commented on compiler/rustc_hir_typeck/src/fallback.rs:
```diff
@@ -464,6 +472,46 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
+                for &diverging_vid in diverging_vids {
+                    let diverging_ty = Ty::new_var(self.tcx, diverging_vid);
+
+                    _ = ocx.eq(&ObligationCause::dummy(), self.param_env, diverging_ty, fallback);
```
We shouldn't expect an error here?

WaffleLapkin replied:
I believe this shouldn't ever error, because we are equating unresolved vars, which should not fail I think?

compiler-errors replied:
pls unwrap then rather than throwing away the result

compiler-errors commented on compiler/rustc_hir_typeck/src/fallback.rs:
```diff
@@ -464,6 +472,46 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
+                for &diverging_vid in diverging_vids {
+                    let diverging_ty = Ty::new_var(self.tcx, diverging_vid);
+
+                    _ = ocx.eq(&ObligationCause::dummy(), self.param_env, diverging_ty, fallback);
```
```suggestion
                    let () = ocx.eq(&ObligationCause::dummy(), self.param_env, diverging_ty, fallback).expect("expected diverging var to be unconstrained");
```

WaffleLapkin replied:
why `let () =`?

compiler-errors replied:
because it doesn't implicitly assume the return type of `eq` is `Result<(), _>`

compiler-errors replied:
either that, or just `ocx.eq().expect` without the `let ()`

WaffleLapkin committed:
Add a fast exit to `lint_obligations_broken_by_never_type_fallback_change`

WaffleLapkin committed:
Apply review nits

WaffleLapkin committed:
Bless a test

WaffleLapkin commented:
Quite a few of existing tests get the warning, should I do something with it? The options I see are bless them (what I currently did), fix them and allow the warning.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9152632872/job/25160381996) [(plain)](https://github.com/rust-lang/rust/commit/fec6f70a33136e3c75205e979d83e50026ee8d18/checks/25160381996/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

     Running tests/compile-test.rs (obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/compile_test-8b497fff66f6b39f)

FAILED TEST: tests/ui/new_ret_no_self.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-900b343f2b3d1665.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-dceec423f9bb6517.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-a1595ad166bef097.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c99e380ec73fb87e.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-ae846d6788d6a9f1.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-283aea41a84e0e95.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-d01c514ff6129ef2.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-facaf493acd2f9ad.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-5689c7175177147d.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-3a5af46fd384dda4.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-283177df69fb203d.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-db42cb36048e8846.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-e9d338b744af34e7.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/new_ret_no_self.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/new_ret_no_self.stderr` to the actual output
--- tests/ui/new_ret_no_self.stderr
+++ <stderr output>
+++ <stderr output>
+error: this function depends on never type fallback being `()`
+   |
+   |
+LL |         fn new(t: T) -> impl Trait2<(), Self> {
+   |
+   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+   = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
+   = help: specify the types explicitly
+   = help: specify the types explicitly
+   = note: `-D dependency-on-unit-never-type-fallback` implied by `-D warnings`
+   = help: to override `-D warnings` add `#[allow(dependency_on_unit_never_type_fallback)]`
+
+error: this function depends on never type fallback being `()`
+   |
+   |
+LL |         fn new(t: T) -> impl Trait2<(), i32> {
+   |
+   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
+   = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
+   = help: specify the types explicitly
+   = help: specify the types explicitly
+
 error: methods called `new` usually return `Self`
... 102 lines skipped ...
    | |_________^
 
-error: aborting due to 12 previous errors
-error: aborting due to 12 previous errors
+error: aborting due to 14 previous errors
 


full stderr:
error: this function depends on never type fallback being `()`
   |
   |
LL |         fn new(t: T) -> impl Trait2<(), Self> {
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
   = help: specify the types explicitly
   = help: specify the types explicitly
   = note: `-D dependency-on-unit-never-type-fallback` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dependency_on_unit_never_type_fallback)]`

error: this function depends on never type fallback being `()`
   |
   |
LL |         fn new(t: T) -> impl Trait2<(), i32> {
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #123748 <https://github.com/rust-lang/rust/issues/123748>
   = help: specify the types explicitly
   = help: specify the types explicitly

error: methods called `new` usually return `Self`
   |
   |
LL | /     pub fn new(_: String) -> impl R<Item = u32> {
LL | |
LL | |         S3
LL | |     }
   | |_____^
   | |_____^
   |
   = note: `-D clippy::new-ret-no-self` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::new_ret_no_self)]`

error: methods called `new` usually return `Self`
   |
LL | /     pub fn new() -> u32 {
LL | |
LL | |         unimplemented!();
LL | |         unimplemented!();
LL | |     }
   | |_____^

error: methods called `new` usually return `Self`
   |
LL | /     pub fn new(_: String) -> u32 {
LL | |
LL | |         unimplemented!();
LL | |         unimplemented!();
LL | |     }
   | |_____^

error: methods called `new` usually return `Self`
   |
   |
LL | /     pub fn new() -> (u32, u32) {
LL | |         unimplemented!();
LL | |     }
   | |_____^


error: methods called `new` usually return `Self`
   |
   |
LL | /     pub fn new() -> *mut V {
LL | |         unimplemented!();
LL | |     }
   | |_____^


error: methods called `new` usually return `Self`
   |
LL | /     pub fn new() -> Option<u32> {
LL | |
LL | |         unimplemented!();
LL | |         unimplemented!();
LL | |     }
   | |_____^

error: methods called `new` usually return `Self`
   |
LL |         fn new() -> String;
   |         ^^^^^^^^^^^^^^^^^^^


error: methods called `new` usually return `Self`
   |
LL |         fn new(_: String) -> String;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^


error: methods called `new` usually return `Self`
   |
   |
LL | /         fn new() -> (u32, u32) {
LL | |             unimplemented!();
LL | |         }
   | |_________^


error: methods called `new` usually return `Self`
   |
   |
LL | /         fn new() -> *mut V {
LL | |             unimplemented!();
LL | |         }
   | |_________^


error: methods called `new` usually return `Self`
   |
   |
LL | /         fn new(t: T) -> impl Into<i32> {
LL | |             1
LL | |         }
   | |_________^


error: methods called `new` usually return `Self`
   |
   |
LL | /         fn new(t: T) -> impl Trait2<(), i32> {
LL | |             unimplemented!()
LL | |         }
   | |_________^

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #125313) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9175045804/job/25227202170) [(plain)](https://github.com/rust-lang/rust/commit/cedada878d60fb049e260e581882d0ee4befb1e1/checks/25227202170/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.9s done
#16 DONE 34.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.51s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
...........i...i..............i.i...............

failures:

---- compiler/rustc_lint_defs/src/builtin.rs - builtin::DEPENDENCY_ON_UNIT_NEVER_TYPE_FALLBACK (line 4343) stdout ----
warning: this function depends on never type fallback being `()`
  |
3 | fn main() {
  | ^^^^^^^^^
  |
---
    compiler/rustc_lint_defs/src/builtin.rs - builtin::DEPENDENCY_ON_UNIT_NEVER_TYPE_FALLBACK (line 4343)

test result: FAILED. 112 passed; 1 failed; 23 ignored; 0 measured; 0 filtered out; finished in 882.86ms

error: doctest failed, to rerun pass `-p rustc_lint_defs --doc`
  local time: Tue May 21 13:28:32 UTC 2024
  network time: Tue, 21 May 2024 13:28:33 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9196777991/job/25295674412) [(plain)](https://github.com/rust-lang/rust/commit/596bbcb5713a724b687b18feac2188abdf91b66c/checks/25295674412/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.1s done
#16 DONE 34.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.36s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.11s
##[endgroup]
##[group]Testing stage0 Linkcheck (x86_64-unknown-linux-gnu)
rustc/lints/listing/warn-by-default.html:1052: broken link - `rustc/lints/listing/prim@never`
rustc/lints/listing/warn-by-default.html:1052: broken link - `rustc/lints/listing/prim@unit`
rustc/print.html:4200: broken link - `rustc/lints/listing/prim@never`
rustc/print.html:4200: broken link - `rustc/lints/listing/prim@unit`
number of HTML files scanned: 38382
number of HTML redirects found: 12093
number of links checked: 3005104
number of links ignored due to external: 70643
```

</details>

WaffleLapkin committed:
Fixup links in lint docs

looks like prim@ stuff does not work here (is it possibly not handled by rustdoc at all?)

compiler-errors commented on src/tools/clippy/tests/ui/new_ret_no_self.rs:
```diff
@@ -73,7 +73,7 @@ struct T;
     // should not trigger lint
     pub fn new() -> Self {
-        unimplemented!();
+        unimplemented!()
```
Wait, why does this have to change? 😅

compiler-errors replied:
Having to change this feels inconsistent with what's in https://github.com/rust-lang/rust/pull/123590#issuecomment-2089207009, but I'm possibly missing something?

WaffleLapkin replied:
it doesn't have to change. I just changed it partly because of preference, partly because this might break on the future. 

(note that comment you linked does accept breakage like this; only saying that changing `return` style is too much)

compiler-errors replied:
Please change this in a different PR then, I'd rather not entangle unnecessary changes into this one

WaffleLapkin replied:
I reverted these changes, so the commit/this PR only include the necessary changes to make this lint not fire for the clippy test

WaffleLapkin committed:
Fixup clippy tests

Don't depend on the fact that `!` falls back to `()` and so panic-ish things
can be used in `-> impl ImplementedForUnit` functions

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit b73eb9af583ecc605339f0de656a72bc657616f3 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors b73eb9af583ecc605339f0de656a72bc657616f3 -->
<!-- homu: {"type":"Approved","sha":"b73eb9af583ecc605339f0de656a72bc657616f3","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
actually wait wait i forgot to run perf ever

@bors r-

compiler-errors commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b73eb9af583ecc605339f0de656a72bc657616f3 with merge 3b63bac3a576b3585ff60fba3a402eb40d159591...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b73eb9af583ecc605339f0de656a72bc657616f3","merge_sha":"3b63bac3a576b3585ff60fba3a402eb40d159591"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9472424932/job/26101093532)
Build commit: 3b63bac3a576b3585ff60fba3a402eb40d159591 (`3b63bac3a576b3585ff60fba3a402eb40d159591`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9472424932/job/26101093532"},"merge_sha":"3b63bac3a576b3585ff60fba3a402eb40d159591"} -->

rust-timer commented:
Queued 3b63bac3a576b3585ff60fba3a402eb40d159591 with parent 3ea5e236ecb4c5f22437059f82d3915d311e4ec0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3ea5e236ecb4c5f22437059f82d3915d311e4ec0&end=3b63bac3a576b3585ff60fba3a402eb40d159591).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (3b63bac3a576b3585ff60fba3a402eb40d159591): [comparison URL](https://perf.rust-lang.org/compare.html?start=3ea5e236ecb4c5f22437059f82d3915d311e4ec0&end=3b63bac3a576b3585ff60fba3a402eb40d159591&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3ea5e236ecb4c5f22437059f82d3915d311e4ec0&end=3b63bac3a576b3585ff60fba3a402eb40d159591&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3ea5e236ecb4c5f22437059f82d3915d311e4ec0&end=3b63bac3a576b3585ff60fba3a402eb40d159591&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=3ea5e236ecb4c5f22437059f82d3915d311e4ec0&end=3b63bac3a576b3585ff60fba3a402eb40d159591&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=3ea5e236ecb4c5f22437059f82d3915d311e4ec0&end=3b63bac3a576b3585ff60fba3a402eb40d159591&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 676.994s -> 679.008s (0.30%)
**Artifact size**: 320.03 MiB -> 320.10 MiB (0.02%)

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit b73eb9af583ecc605339f0de656a72bc657616f3 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors b73eb9af583ecc605339f0de656a72bc657616f3 -->
<!-- homu: {"type":"Approved","sha":"b73eb9af583ecc605339f0de656a72bc657616f3","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b73eb9af583ecc605339f0de656a72bc657616f3 with merge a3baa7b185fafa628de74905d3b768feb6ad3fab...
<!-- homu: {"type":"BuildStarted","head_sha":"b73eb9af583ecc605339f0de656a72bc657616f3","merge_sha":"a3baa7b185fafa628de74905d3b768feb6ad3fab"} -->

rust-log-analyzer commented:
The job **`rfl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9476224580/job/26108794466) [(plain)](https://github.com/rust-lang-ci/rust/commit/a3baa7b185fafa628de74905d3b768feb6ad3fab/checks/26108794466/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_traits test:false 13.634
   Compiling rustc_mir_transform v0.0.0 (/checkout/compiler/rustc_mir_transform)
[RUSTC-TIMING] rustc_resolve test:false 66.318
   Compiling rustc_hir_typeck v0.0.0 (/checkout/compiler/rustc_hir_typeck)
error[E0599]: no variant or associated item named `FallbackToUnit` found for enum `DivergingFallbackBehavior` in the current scope
    |
    |
20  | pub enum DivergingFallbackBehavior {
    | ---------------------------------- variant or associated item `FallbackToUnit` not found for this enum
...
480 |         let DivergingFallbackBehavior::FallbackToUnit = behavior else { return };
    |                                        ^^^^^^^^^^^^^^ variant or associated item not found in `DivergingFallbackBehavior`
[RUSTC-TIMING] rustc_mir_build test:false 43.190
   Compiling rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
For more information about this error, try `rustc --explain E0599`.
[RUSTC-TIMING] rustc_hir_typeck test:false 5.945
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9476224580/job/26108981225)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9476224580/job/26108981225","builder_name":"checks-actions"} -->

cjgillot commented on tests/ui/delegation/not-supported.rs:
```diff
@@ -70,12 +70,16 @@ mod opaque {
         pub fn opaque_arg(_: impl Trait) -> i32 { 0 }
         pub fn opaque_ret() -> impl Trait { unimplemented!() }
+        //~^ warn: this function depends on never type fallback being `()`
+        //~| warn: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
```
Is there a way to show what error would be emitted?

compiler-errors replied:
I've implemented a rudimentary attempt at https://github.com/rust-lang/rust/pull/126367

compiler-errors commented:
@bors r-

WaffleLapkin committed:
rebase blessing

bors commented:
:pushpin: Commit ea98e42bfd7e3723a68cb2dd6f6ac4a4866a400e has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors ea98e42bfd7e3723a68cb2dd6f6ac4a4866a400e -->
<!-- homu: {"type":"Approved","sha":"ea98e42bfd7e3723a68cb2dd6f6ac4a4866a400e","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ea98e42bfd7e3723a68cb2dd6f6ac4a4866a400e with merge b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc...
<!-- homu: {"type":"BuildStarted","head_sha":"ea98e42bfd7e3723a68cb2dd6f6ac4a4866a400e","merge_sha":"b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9501134701/job/26192858841)
Approved by: compiler-errors
Pushing b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9501134701/job/26192858841"},"merge_sha":"b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc"} -->

rust-timer commented:
Finished benchmarking commit (b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc): [comparison URL](https://perf.rust-lang.org/compare.html?start=9fdbfe1441a82e96e3f99ae199ff5348d8d948ae&end=b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9fdbfe1441a82e96e3f99ae199ff5348d8d948ae&end=b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.2%, 0.2%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9fdbfe1441a82e96e3f99ae199ff5348d8d948ae&end=b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc&stat=max-rss)
<details>
<summary>Results (primary 1.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.5%  | [5.5%, 5.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-1.2%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.3%  | [-1.2%, 5.5%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9fdbfe1441a82e96e3f99ae199ff5348d8d948ae&end=b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc&stat=cycles:u)
<details>
<summary>Results (primary -3.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.4% | [-7.0%, -1.0%] | 14    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.4% | [-7.0%, -1.0%] | 14    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9fdbfe1441a82e96e3f99ae199ff5348d8d948ae&end=b6e5e3ffbbc41a7ef1f2a522fc7ded19b99918dc&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 672.787s -> 671.147s (-0.24%)
**Artifact size**: 320.34 MiB -> 320.41 MiB (0.02%)
